In [1]:
import requests
import json

In [2]:
with open("secrets/lastfm.json") as f:
    client_json = json.load(f)

In [ ]:
base_url = "http://ws.audioscrobbler.com/2.0/"

# test to see if we get response
headers = {
    "user-agent": "python-tester"
}

payload = {
    "api_key": client_json["client_id"],
    "method": "chart.gettopartists",
    "format": "json"
}

r = requests.get(base_url, headers=headers, params=payload)
r.status_code

200

In [4]:
# test to get auth token
auth_payload = {
    "api_key": client_json["client_id"],
    "method": "auth.gettoken",
    "format": "json"
}

auth_r = requests.get(base_url, headers=headers, params=auth_payload)
auth_r.status_code

200

In [5]:
auth_token = auth_r.json()["token"]
print(auth_token)

RPkvErMFrnZs5GDjNyy6sJwWk-hZ5lvS


In [6]:
# authenticate user?
import webbrowser

auth_request_url = "http://www.last.fm/api/auth"
request_auth_payload = {
    "api_key": client_json["client_id"],
    "token": auth_token
}

request_auth_r = requests.get(auth_request_url, headers=headers, params=request_auth_payload)
request_auth_r.status_code

webbrowser.open(request_auth_r.url)

True

In [8]:
# signing calls, needed for use of services
import hashlib
auth_payload = {
    "api_key": client_json["client_id"],
    "method": "auth.gettoken",
    "format": "json"
}
request_params_signing = "".join([str(k)+str(v) for k, v in auth_payload.items()][:-1]) + client_json["client_secret"]
print(request_params_signing)
print(hashlib.md5(request_params_signing.encode()))

# signing needs
    # - api key
    # - api secret
    # - method
# join "api_key"+api_key-value, "method"+method-string and api_secret

api_key96e8b6dd8a4702e9e058ed121094987fmethodauth.gettoken7f65cc958a0fe5800a535207c43fbcc9
<md5 _hashlib.HASH object @ 0x105da9bb0>


In [9]:
[str(k)+str(v) for k, v in auth_payload.items()][:-1]

['api_key96e8b6dd8a4702e9e058ed121094987f', 'methodauth.gettoken']

lol getRecentTracks trenger ikke auth.

In [18]:
import pytz
from datetime import datetime, timedelta
import webbrowser
import requests
today = pytz.utc.localize(datetime.now().replace(hour=0, minute=0, second=0, microsecond=0))
yesterday = today - timedelta(days=5)
today_ms = int(today.timestamp())
yesterday_ms = int(yesterday.timestamp())

base_url = "http://ws.audioscrobbler.com/2.0/"

headers = {
    "user-agent": "GP-personal-app"
}

recent_tracks_params = {
    "limit": 20,
    "user": "GammelPerson",
    "from": yesterday_ms,
    "to": today_ms,
    "extended": 0,
    "api_key": client_json["client_id"],
    "format": "json",
    "method": "user.getrecenttracks",
}

r = requests.get(base_url, headers=headers, params=recent_tracks_params)
webbrowser.open(r.url)
r.status_code

200

In [17]:
r.url

'http://ws.audioscrobbler.com/2.0/?limit=20&user=GammelPerson&from=1764115200&to=1764547200&extended=1&api_key=96e8b6dd8a4702e9e058ed121094987f&format=json&method=user.getrecenttracks'

In [8]:
date_format = "%Y-%m-%d"
end = "2025-12-01"
int(
        pytz.utc.localize(datetime.strptime(end, date_format)).timestamp()
    )


1764547200

In [6]:

webbrowser.open(r.url)


True

Response er en .json-fil. Den har to nøkler på høyeste nivå, `recent_tracks` og `@attr`. Gitt at du gjør et søk over en lenger periode, for eksempel her med flere dagers mellomrom, er det mulig å få antall sider med resultater i `@attr`. I `recent_tracks.tracks` finner man da all sanginfoen.

Det går vel an å bruke dette for å hente hele lyttehistorien min fra 01.01-30.11. Bare at jeg kanskje burde indeksere sidene og dumpe resultatene til json. Vet ikke om dette overskrider rate_limit satt av last.fm.

In [35]:
r.url

'http://ws.audioscrobbler.com/2.0/?limit=20&user=GammelPerson&from=1764288000000&to=1764374400000&extended=1&api_key=96e8b6dd8a4702e9e058ed121094987f&format=json&method=user.getrecenttracks'

In [11]:
q = r.json()

In [12]:
q

{'recenttracks': {'track': [{'artist': {'mbid': '', '#text': 'Slimesito'},
    'streamable': '0',
    'image': [{'size': 'small',
      '#text': 'https://lastfm.freetls.fastly.net/i/u/34s/a9198665ef7ca70d920e25c509d322ba.jpg'},
     {'size': 'medium',
      '#text': 'https://lastfm.freetls.fastly.net/i/u/64s/a9198665ef7ca70d920e25c509d322ba.jpg'},
     {'size': 'large',
      '#text': 'https://lastfm.freetls.fastly.net/i/u/174s/a9198665ef7ca70d920e25c509d322ba.jpg'},
     {'size': 'extralarge',
      '#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/a9198665ef7ca70d920e25c509d322ba.jpg'}],
    'mbid': '',
    'album': {'mbid': '', '#text': 'Pray 4 Me'},
    'name': 'Pray 4 Me',
    'url': 'https://www.last.fm/music/Slimesito/_/Pray+4+Me',
    'date': {'uts': '1764531747', '#text': '30 Nov 2025, 19:42'}},
   {'artist': {'mbid': 'ec29539f-4e72-46af-9fb0-b5697a3cf41b',
     '#text': 'Evita Manji'},
    'streamable': '0',
    'image': [{'size': 'small',
      '#text': 'https://lastfm

In [14]:
q["recenttracks"]["track"]

[{'artist': {'mbid': '', '#text': 'Slimesito'},
  'streamable': '0',
  'image': [{'size': 'small',
    '#text': 'https://lastfm.freetls.fastly.net/i/u/34s/a9198665ef7ca70d920e25c509d322ba.jpg'},
   {'size': 'medium',
    '#text': 'https://lastfm.freetls.fastly.net/i/u/64s/a9198665ef7ca70d920e25c509d322ba.jpg'},
   {'size': 'large',
    '#text': 'https://lastfm.freetls.fastly.net/i/u/174s/a9198665ef7ca70d920e25c509d322ba.jpg'},
   {'size': 'extralarge',
    '#text': 'https://lastfm.freetls.fastly.net/i/u/300x300/a9198665ef7ca70d920e25c509d322ba.jpg'}],
  'mbid': '',
  'album': {'mbid': '', '#text': 'Pray 4 Me'},
  'name': 'Pray 4 Me',
  'url': 'https://www.last.fm/music/Slimesito/_/Pray+4+Me',
  'date': {'uts': '1764531747', '#text': '30 Nov 2025, 19:42'}},
 {'artist': {'mbid': 'ec29539f-4e72-46af-9fb0-b5697a3cf41b',
   '#text': 'Evita Manji'},
  'streamable': '0',
  'image': [{'size': 'small',
    '#text': 'https://lastfm.freetls.fastly.net/i/u/34s/49d913353ae9b30e4df493b16613274a.jpg'

In [23]:
pytz.utc.localize(datetime(year=2025, month=1, day=1, hour=0, minute=0, second=0))

datetime.datetime(2025, 1, 1, 0, 0, tzinfo=<UTC>)

In [ ]:
# Sjekke hvor mange .json filer jeg må dumpe hvis vi henter 50 sanger per request.

import pytz
from datetime import datetime, timedelta
end = pytz.utc.localize(datetime(year=2025, month=12, day=1, hour=0, minute=0, second=0))
start = pytz.utc.localize(datetime(year=2025, month=1, day=1, hour=0, minute=0, second=0))
start_ms = int(start.timestamp())
end_ms = int(end.timestamp())

base_url = "http://ws.audioscrobbler.com/2.0/"

recent_tracks_params = {
    "limit": 50,
    "user": "GammelPerson",
    "from": start_ms,
    "to": end_ms,
    "extended": 0,
    "api_key": client_json["client_id"],
    "format": "json",
    "method": "user.getrecenttracks",
    "page": 1,
}

r = requests.get(base_url, headers=headers, params=recent_tracks_params)
webbrowser.open(r.url)
r.status_code

print(r.json()["recenttracks"]["@attr"])

{'user': 'GammelPerson', 'totalPages': '321', 'page': '1', 'perPage': '50', 'total': '16022'}


I perioden 01.01-30.11 blir det totalt 321 sider, eller 321 .json-filer totalt.

Kan lage en cron-jobb som tar og samler inn data fra foregående dag.